Conversions between slices and boxes

clarcharr opened the pull request:
This allows conversion for `Copy` slices, `str`, and `CStr` into their boxed counterparts.

This also adds the method `CString::into_boxed_c_str`.

I would like to add similar implementations for `OsStr` as well, but I have not figured out how.

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

aturon commented:
Since these impls are instantly stable, we need the full libs team to sign off.

Personally, I think that we will continue growing such impls until the end of time. Due to trait coherence, `std` is the only place such impls can live. So I'm fine continuing to sign off on such PRs.

@rfcbot fcp merge

aturon commented:
Whoops, didn't have the libs team assigned first, trying again:

@rfcbot fcp merge

clarcharr commented:
Reminder to self for later: also add versions for CStr and OsStr

alexcrichton commented:
@clarcharr the compile failures here seem worrisome, they may want to be fixed!

clarcharr commented:
@alexchricton I'll fix those soon; I know what's wrong.

Instead of using `Vec` and `String` I'll have to do it manually.

clarcharr commented:
Waiting on tests, but this should work now. I also added an implementation for `CStr` and added a new method, `CString::into_boxed_c_str`, which will need a tracking issue once I get an informal nod that such a method is okay to add.

alexcrichton commented:
@clarcharr perhaps the `CStr` parts could be left out for a future PR? Unfortunately trait impls are instantly stable today and we don't have much other precedent with that type and DST impls, so we may just wish to conservatively tackle it in a separate PR

clarcharr commented:
Sounds good to me! I'll split them out and open another PR.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 550373b has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 550373b1f9d59a4ce5cd61cb1bca2488c0ec6471 -->

malbarbo commented:
Isn't `T: Copy` to restrictive? How about `T: Clone`?

clarcharr commented:
@malbarbo That's because it's more efficient if the type implements `Copy`, but I could make a more general version using impl specialisation. If you think it's worth having, feel free to open an issue and I might get around to it!

malbarbo commented:
@clarcharr You can use `Vec::from_iter`, it is already specialized for copy types https://doc.rust-lang.org/stable/src/collections/vec.rs.html#1499-1670

clarcharr commented:
@malbarbo right now that'll work for `.to_vec().into_boxed_slice()`, but not directly like in these conversions. I'll take a look at it in my other PR.

clarcharr commented:
So apparently this didn't get merged? Did I miss something?

bors commented:
:hourglass: Testing commit 550373b with merge 8482c4b...

bjorn3 commented:
@clarcharr The rollup pr failed, so you will have to wait till the next rollup. (Got a pr of mine merged after 6 rollup prs)

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1883)

alexcrichton commented:
@bors: retry

* #38620

On Fri, Feb 10, 2017 at 12:40 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-appveyor
> <https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1883>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39438#issuecomment-278887818>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95C5k6K2gP7Th6wU9oi22LPWVIz9Kks5rbCKFgaJpZM4LzdpA>
> .
>

bors commented:
:hourglass: Testing commit 550373b with merge 16f5492...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200419816)

bjorn3 commented:
dist-mips-linux:
```
[...]

[ 83%] Built target LLVMMSP430Info
Scanning dependencies of target LLVMMSP430Desc
[ 83%] Building CXX object lib/Target/MSP430/MCTargetDesc/CMakeFiles/LLVMMSP430Desc.dir/MSP430MCTargetDesc.cpp.o


No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

alexcrichton commented:
@bors: retry

* sccache wedged #39003

On Fri, Feb 10, 2017 at 11:18 AM, bjorn3 <notifications@github.com> wrote:

> dist-mips-linux:
>
> [...]
>
> [ 83%] Built target LLVMMSP430Info
> Scanning dependencies of target LLVMMSP430Desc
> [ 83%] Building CXX object lib/Target/MSP430/MCTargetDesc/CMakeFiles/LLVMMSP430Desc.dir/MSP430MCTargetDesc.cpp.o
>
>
> No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
> Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
>
> The build has been terminated
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39438#issuecomment-279039468>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95D926SXesz_zCfgWxi0pvlElq3ccks5rbLgdgaJpZM4LzdpA>
> .
>

bors commented:
:hourglass: Testing commit 550373b with merge 99b6677...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200471781)

bors commented:
:hourglass: Testing commit 550373b with merge 2425b22...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1895), [status-travis](https://travis-ci.org/rust-lang/rust/builds/200493369)
Approved by: alexcrichton
Pushing 2425b227742bb9214d60429d1325816e0b93e152 to master...

japaric commented:
@alexcrichton Sadly, no. I haven't seen the thread sanitizer not catch that data race before. I'll test locally to see if I can repro.
